044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,353 @@
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.soft=0.05`
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.hard=0.01`
+
+Or mixing the two types, which is also valid to soft limit at 0.05 (5%) availableRatio and hard limit at 1GB availableBytes:
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.soft=0.05`
+- `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.hard=1000000000`
+
+Defining more than one hard or soft limit is **invalid**, for example the below is disallowed:
+- `client.quota.callback.static.storage.perVolumeLimit.availableRatioBelow.hard=0.05`
+- `client.quota.callback.static.storage.perVolumeLimit.availableBytesBelow.hard=5000000000`
+
+#### Throttle Factor Fallback
+
+We are going to depend on using the admin client to get volume information. This brings all the baggage of making a
+connection and dealing with possible failures. Also, we can potentially have an inconsistent view of the world between
+determining the active broker set and asking for a description of all log dirs (because we make two independent API calls)
+. So we need to react somehow to cases where we cannot get the volume data for all active brokers.
+
+Example inconsistent state:
+1. we call `describeCluster` and get a response that says broker 1 and 2 are active
+2. broker 2 shuts down cleanly and is removed from the active set
+3. we call `describeLogDirs( brokerIds = [1,2] )` and only receive descriptions for logdirs on broker 1
+
+We propose introducing a configurable **throttle factor fallback** to be applied in situations where we don't have enough
+information to act. With a default value of 1.0 to optimistically allow all the quota to be used. This would allow
+users to opt in to more pessimistic behaviour like using a factor of 0.0 to prevent writes when we are in an unknown
+state.
+
+Throttle factor fallback can be in the range (0.0, 1.0)
+
+Example configuration: `client.quota.callback.static.throttle.factor.fallback=0.0`
+
+##### Throttle Factor Valid For Polls
+
+To make the throttle factor logic more resilient in the face of transient failures we want to continue
+using a previously calculated throttle factor for some duration, rather than immediately using the throttle factor
+fallback after a single failure.
+
+So we will enable the user to specify how many poll intervals a valid throttle factor should apply for with
+`client.quota.callback.static.throttle.factor.valid-for-polls`. Default value 2.
+
+For example if validity intervals is 1 then we could expect this behaviour:
+
+1. poll #1 success, calculated throttle factor 0.6
+2. poll #2 failed to calculate throttle factor, keep using 0.6 as we are <= 1 interval after 0.6 was calculated
+3. poll #3 failed, use the throttle factor fallback of 1.0 because we are > 1 interval after 0.6 was calculated
```

Comment:
AFAIU we are talking about failures on getting cluster information (disk usage). Of course it's doing in a "polling" way but at same time "poll" is related to consuming messages. Should we clarify maybe not talking about poll and using another name for the `valid-for-polls` parameter? I was wondering if it could confuse the user.

Reply:
We've changed it to a `validity-duration` to decouple it from this idea of 'polling/recalculation'. I think it's better this way as it doesn't vary as you change the recalculation interval, less surprising.